sycl/source/detail/program_impl.cpp

```diff
@@ -146,21 +146,7 @@ program_impl::program_impl(ContextImplPtr Context,
                                             sizeof(sycl::detail::pi::PiDevice) *
                                                 NumDevices,
                                             PiDevices.data(), nullptr);
-  std::vector<device> SyclContextDevices =
-      MContext->get_info<info::context::devices>();
-
-  // Keep only the subset of the devices (associated with context) that
-  // were actually used to create the program.
-  // This is possible when clCreateProgramWithBinary is used.
-  auto NewEnd = std::remove_if(
-      SyclContextDevices.begin(), SyclContextDevices.end(),
-      [&PiDevices](const sycl::device &Dev) {
-        return PiDevices.end() ==
-               std::find(PiDevices.begin(), PiDevices.end(),
-                         detail::getSyclObjImpl(Dev)->getHandleRef());
-      });
-  SyclContextDevices.erase(NewEnd, SyclContextDevices.end());
-  MDevices = SyclContextDevices;
+  MDevices = MContext->get_info<info::context::devices>();
```

Comment:
Above this line, we retrieve devices this program has been created with, but with this patch we just essentially dismiss that information and use all devices from the context. That seems incorrect, for example, if the program was only created for a single device from a multi-device context. In that case we might end up querying the wrong device on L155.

Do we even need to retrieve `sycl::device`s from the context here? It seems that we're mostly working with `PiDevice`s in `program_impl`. `program_impl::get_devices` is an exception that comes to mind (since it returns MDevices) but I'm not sure if we actually need it anywhere, considering that `program_impl` is now purely an internal class.

Reply:
> The current filtering is, of course, incorrect now that subdevices can be used within contexts that contain their parent devices. But do we even need to retrieve `sycl::device`s from the context here?

Yes, the added test case fails because of the incorrect filtering code.
The filtering code removes the subdevice which created the context and ends up "no device" error.

Reply:
Right. What I'm saying is that it makes more sense to use PiDevices retrieved on the line above with piProgramGetInfo instead of just assuming that the program was created with all devices in the passed context and using PiDevices retrieved from those sycl::devices, which is what your patch does.

For example, ff the program was created with only the second device in the context, then the call to piProgramGetBuildInfo on L155 will be made with the wrong device.

Reply:
MDevices should only contain device_impls associated with PiDevices fetched earlier. Your code right now fetches all of the devices in that platform. Instead you should just fetch device_impl for each pi_device in PiDevices using platform_impl::getDeviceImpl function.